sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -327,23 +353,48 @@ private[joins] object UnsafeHashedRelation {
     // Create a mapping of buildKeys -> rows
     val keyGenerator = UnsafeProjection.create(key)
     var numFields = 0
-    while (input.hasNext) {
-      val row = input.next().asInstanceOf[UnsafeRow]
-      numFields = row.numFields()
-      val key = keyGenerator(row)
-      if (!key.anyNull) {
+
+    if (isLookupAware) {
+      // Add one extra boolean value at the end as part of the row,
+      // to track the information that whether the corresponding key
+      // has been looked up or not. See `ShuffledHashJoin.fullOuterJoin` for example of usage.
+      val valueGenerator = UnsafeProjection.create(value.get :+ Literal(false))
+
+      while (input.hasNext) {
+        val row = input.next().asInstanceOf[UnsafeRow]
+        numFields = row.numFields() + 1
+        val key = keyGenerator(row)
+        val value = valueGenerator(row)
         val loc = binaryMap.lookup(key.getBaseObject, key.getBaseOffset, key.getSizeInBytes)
         val success = loc.append(
           key.getBaseObject, key.getBaseOffset, key.getSizeInBytes,
-          row.getBaseObject, row.getBaseOffset, row.getSizeInBytes)
+          value.getBaseObject, value.getBaseOffset, value.getSizeInBytes)
         if (!success) {
           binaryMap.free()
           // scalastyle:off throwerror
           throw new SparkOutOfMemoryError("There is not enough memory to build hash map")
           // scalastyle:on throwerror
         }
-      } else if (isNullAware) {
-        return EmptyHashedRelationWithAllNullKeys
+      }
+    } else {
+      while (input.hasNext) {
+        val row = input.next().asInstanceOf[UnsafeRow]
+        numFields = row.numFields()
+        val key = keyGenerator(row)
+        if (!key.anyNull) {
+          val loc = binaryMap.lookup(key.getBaseObject, key.getBaseOffset, key.getSizeInBytes)
+          val success = loc.append(
+            key.getBaseObject, key.getBaseOffset, key.getSizeInBytes,
+            row.getBaseObject, row.getBaseOffset, row.getSizeInBytes)
+          if (!success) {
+            binaryMap.free()
+            // scalastyle:off throwerror
+            throw new SparkOutOfMemoryError("There is not enough memory to build hash map")
+            // scalastyle:on throwerror
+          }
+        } else if (isNullAware) {
+          return EmptyHashedRelationWithAllNullKeys
+        }
```

Comment:
nit format: how about reformatting it like this?
```

    val append = (key: UnsafeRow, value: UnsafeRow) => {
      val loc = binaryMap.lookup(key.getBaseObject, key.getBaseOffset, key.getSizeInBytes)
      val success = loc.append(
        key.getBaseObject, key.getBaseOffset, key.getSizeInBytes,
        value.getBaseObject, value.getBaseOffset, value.getSizeInBytes)
      if (!success) {
        binaryMap.free()
        // scalastyle:off throwerror
        throw new SparkOutOfMemoryError("There is not enough memory to build hash map")
        // scalastyle:on throwerror
      }
    }

    if (isLookupAware) {
      // Add one extra boolean value at the end as part of the row,
      // to track the information that whether the corresponding key
      // has been looked up or not. See `ShuffledHashJoin.fullOuterJoin` for example of usage.
      val valueGenerator = UnsafeProjection.create(value.get :+ Literal(false))
      while (input.hasNext) {
        val row = input.next().asInstanceOf[UnsafeRow]
        numFields = row.numFields() + 1
        append(keyGenerator(row), valueGenerator(row))
      }
    } else {
      while (input.hasNext) {
        val row = input.next().asInstanceOf[UnsafeRow]
        numFields = row.numFields()
        val key = keyGenerator(row)
        if (!key.anyNull) {
          append(key, row)
        } else if (isNullAware) {
          return EmptyHashedRelationWithAllNullKeys
        }
      }
    }
```

Reply:
+1

Reply:
@maropu - thanks, it looks better, updated.